test-suite/barrieroption.cpp

```diff
@@ -246,6 +246,100 @@ BOOST_AUTO_TEST_CASE(testParity) {
                     << "\n    error:      " << error);
     }
 }
+
+BOOST_AUTO_TEST_CASE(testPutCallSymmetry) {
+    BOOST_TEST_MESSAGE(
+        "Testing put-call symmetry for barrier options...");
+
+    Date today = Settings::instance().evaluationDate();
+
+    struct PutCallSymmetryTestCase {
+        Real callStrike;
+        Real callBarrier;
+        Barrier::Type callType;
+        Real putStrike;
+        Real putBarrier;
+        Barrier::Type putType;
+    };
+
+    PutCallSymmetryTestCase cases[] = {
+        { 90, 95, Barrier::DownOut, 111.11111, 105.26315, Barrier::UpOut },
+        { 95, 95, Barrier::DownOut, 105.26315, 105.26315, Barrier::UpOut },
+        { 100, 95, Barrier::DownOut, 100.0, 105.26315, Barrier::UpOut },
+        { 105, 95, Barrier::DownOut, 95.23809, 105.26315, Barrier::UpOut },
+        { 110, 95, Barrier::DownOut, 90.90909, 105.26315, Barrier::UpOut },
+
+        { 90.0, 120.0, Barrier::UpOut, 111.11111, 83.33333, Barrier::DownOut },
+        { 95.0, 120.0, Barrier::UpOut, 105.26315, 83.33333, Barrier::DownOut },
+        { 100.0, 120.0, Barrier::UpOut, 100.0, 83.33333, Barrier::DownOut },
+        { 105.0, 120.0, Barrier::UpOut, 95.23809, 83.33333, Barrier::DownOut },
+        { 110.0, 120.0, Barrier::UpOut, 90.90909, 83.33333, Barrier::DownOut }
+    };
+
+    DayCounter dc = Actual360();
+    Date maturity = today + 360;
+    ext::shared_ptr<Exercise> exercise =
+        ext::make_shared<EuropeanExercise>(maturity);
+    Real r = 0.01;
+    Real rebate = 0.0;
+    Real spotPrice = 100;
+
+    ext::shared_ptr<SimpleQuote> spot = ext::make_shared<SimpleQuote>();
```

Comment:
These declarations can all be shortened to 
```
    auto spot = ext::make_shared<SimpleQuote>();
```
The same goes for the processes, engines and payoffs below.